### PR TITLE
Add namespace

### PR DESCRIPTION
**Pull request by alex-lairan**
### Description of the Change

Add a namespace DSL to the Router DSL.

The namespace act as a local scope.

This DSL result to 
```
get "/hello"
namespace "/world" do
  get "/foo"
end
```

```
get "/hello"
get "/world/foo"
```

### Alternate Designs

Other keywords can be used instead of `namespace`, like `scope`.

### Benefits

This change will reduce repetions.

### Possible Drawbacks

None.

### CODE REVIEW

File: src/amber/dsl/router.cr [hunk 1]
```diff
@@ -25,13 +25,16 @@ module Amber::DSL
...
+      {{with DSL::Router.new(self, valve, scope + scoped_namespace) yield}}
+    end
+
     {% for verb in RESOURCES %}
       macro {{verb.id}}(*args)
         route {{verb}}, \{{*args}}
         {% if verb == :get %}
         route :head, \{{*args}}
         {% end %}
-        route {{verb}}, \{{*args}}
```
Comment (drujensen) [inline src/amber/dsl/router.cr]: hmm, why was this added twice? nice catch!

File: spec/amber/dsl/server_spec.cr [hunk 2]
```diff
@@ -124,6 +124,38 @@ module Amber
...
+          Amber::Server.configure do
+            pipeline :web do
+            end
+
+            routes :web do
+              namespace "/test" do
+                get "/hello", HelloController, :index
+
+                namespace "/test2" do
+                  get "/hello", HelloController, :index
```
Comment (drujensen) [inline spec/amber/dsl/server_spec.cr]: Can you add a `resource` to a namespace?

File: src/amber/dsl/router.cr [hunk 3]
```diff
@@ -25,13 +25,16 @@ module Amber::DSL
       router.add(%route)
     end
 
+    macro namespace(scoped_namespace)
+      {{with DSL::Router.new(self, valve, scope + scoped_namespace) yield}}
```
Comment (drujensen) [inline src/amber/dsl/router.cr]: Clever.  💯

File: spec/amber/dsl/server_spec.cr (hunk 2 above)
Comment (alex-lairan) [reply spec/amber/dsl/server_spec.cr]: `resource` is in the same namespace as `get` `post` etc, I think so.

File: src/amber/dsl/router.cr (hunk 3 above)
Comment (alex-lairan) [reply src/amber/dsl/router.cr]: Thanks !

But it not work :( 

The scope don't change.

Comment (alex-lairan) [reply src/amber/dsl/router.cr]: I tried 
```
      old_scope = scope
      new_scope = scope + {{scoped_namespace}}

      scope = new_scope
      {{yield}}
      scope = old_scope
```

But I got 

```
instantiating 'Amber::Router::Router#draw(Symbol, String)'
in spec/amber/dsl/server_spec.cr:138: macro didn't expand to a valid program, it expanded to:

================================================================================
--------------------------------------------------------------------------------
   1.       puts "scope before #{scope}"
   2.       old_scope = scope
   3.       new_scope = scope + "/test"
   4. 
   5.       scope = new_scope
   6.       begin get("/hello", HelloController, :index)
   7. namespace("/test2") do
   8.   get("/hello", HelloController, :index)
   9. end
  10.  end
  11.       scope = old_scope
  12. 
  13.       puts "scope after #{new_scope}"
  14.       # begin get("/hello", HelloController, :index)
  15. namespace("/test2") do
  16.   get("/hello", HelloController, :index)
  17. end
  18.  end
  19.     
--------------------------------------------------------------------------------
Syntax error in spec/amber/dsl/server_spec.cr:144: expecting token 'EOF', not 'end'

              end
 ^

================================================================================

              namespace "/test" do
              ^~~~~~~~~
```

File: src/amber/dsl/router.cr (hunk 1 above)
Comment (alex-lairan) [reply src/amber/dsl/router.cr]: Maybe a rebase or a merge conflict :smile:

**Comment by alex-lairan** [comment]
I found a way by adding scope handle inside the router.

But I don't like the implementation, anyone have an idea ?

Another problem, `{{yield}}` is not working, I need to `with ... yield`.

**Comment by alex-lairan** [comment]
I tried on my app, it works !

File: spec/amber/dsl/server_spec.cr (hunk 2 above)
Comment (robacarp) [reply spec/amber/dsl/server_spec.cr]: @alex-lairan can you add a test for a resource inside the namespace?

File: src/amber/router/router.cr [hunk 4]
```diff
@@ -11,6 +11,19 @@ module Amber
...
+
+      def push_scope(scope : String)
+        @scopes << scope
+      end
+
+      def pop_scope
+        @scopes.pop
+      end
+
+      private def actual_scope
```
Comment (robacarp) [inline src/amber/router/router.cr]: This probably doesn't need the `actual_` prefix, a private getter for just `scope` makes sense. The prefix `actual_` makes me think there'd be another method lying around called `fake_scope` or something.

File: src/amber/router/router.cr [hunk 5]
```diff
@@ -11,6 +11,19 @@ module Amber
         @routes = RouteSet(Route).new
         @routes_hash = {} of String => Route
         @socket_routes = Array(NamedTuple(path: String, handler: WebSockets::Server::Handler)).new
+        @scopes = [] of String
+      end
+
+      def push_scope(scope : String)
+        @scopes << scope
+      end
```
Comment (robacarp) [inline src/amber/router/router.cr]: Do we have a router_spec where we can add tests for these methods?

File: src/amber/router/router.cr [hunk 6]
```diff
@@ -11,6 +11,19 @@ module Amber
...
+
+      def pop_scope
+        @scopes.pop
+      end
```
[the remark this answers is not in the document]
Comment (alex-lairan) [reply src/amber/router/router.cr]: Thanks, I improved the implementation on it's own class.

**Comment by alex-lairan** [comment]
Add a Scope class.

This will simplify the code and tests ! :)

File: spec/amber/dsl/server_spec.cr (hunk 2 above)
Comment (alex-lairan) [reply spec/amber/dsl/server_spec.cr]: Yes, no problem !

Comment (alex-lairan) [reply spec/amber/dsl/server_spec.cr]: Done

File: src/amber/router/scope.cr [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+      def push(scope : String) : Array(String)
+        @stack << scope
+      end
+
+      def pop : String
+        @stack.pop
+      end
+
+      def to_s(io)
+        io << @stack.join("")
```
[the remark this answers is not in the document]
Comment (alex-lairan) [reply src/amber/router/scope.cr]: Done :+1:

File: spec/amber/dsl/server_spec.cr [hunk 8]
```diff
@@ -124,6 +124,45 @@ module Amber
...
+      describe "routes" do
+        context "with a namespace" do
+          server = Amber::Server.instance
+          server.handler.pipeline.clear
+
+          Amber::Server.configure do
+            pipeline :web do
+            end
+
+            routes :web do
```
Comment (drujensen) [inline spec/amber/dsl/server_spec.cr]: Can you add one more test?  You can add a namespace to a `routes` as follows:
```
routes :api, "/api" do
  namespace "/test" do
    resources "/user", UserController
  end
end
```
Can you verify this creates `/api/test/user`?

Comment (alex-lairan) [reply spec/amber/dsl/server_spec.cr]: Done :+1:

**Comment by alex-lairan** [comment]
@drujensen can you re-review please ? :smile:
